File: framework/ui-elements/src/displays/components/Template.js
import AbstractTemplate from "../../../gen/displays/components/AbstractTemplate";
import DisplayFactory from "alexandria-ui-elements/src/displays/DisplayFactory";

export default class Template extends AbstractTemplate {

	constructor(props) {
		super(props);
		this.state = {
		    ...this.state,
		    canClose : true
		};
	};

	enableCloseManager = () => {
	    const template = this;
	    window.addEventListener("beforeunload", function(event) {
	        if (template.state.canClose) return;
            event.returnValue = true;
            return true;
        });
	};

	canClose = (value) => {
	    this.setState({ canClose : value });
	};

	fixHeight = (height) => {
	    // A template may be rendered inside a dialog. Keep percentage heights relative
	    // to its immediate container instead of forcing the viewport height.
	    return height;
	};

}

DisplayFactory.register("Template", Template);

File: framework/ui/src/io/intino/alexandria/ui/displays/components/Template.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.ui.displays.Component;
import io.intino.alexandria.ui.displays.notifiers.TemplateNotifier;

public class Template<DN extends TemplateNotifier, Item, B extends Box> extends Component<DN, B> {
    private Item item;
    private boolean closeManagerEnabled = false;

    public Template(B box) {
        super(box);
    }

    public void value(Item item) {
        item(item);
    }

    public void item(Item item) {
        _item(item);
        refresh();
    }

    public Item item() {
        return item;
    }

    public void canClose(boolean value) {
        enableCloseManager();
        notifier.canClose(value);
    }

    protected Template<DN, Item, B> _item(Item item) {
        this.item = item;
        return this;
    }

    private void enableCloseManager() {
        if (closeManagerEnabled) return;
        notifier.enableCloseManager();
        closeManagerEnabled = true;
    }

}